### PR TITLE
Add Cloud Search

### PR DESCRIPTION
**Pull request by mschae**
This adds cloud search.

There are currently still a few things to do:

- [x] Add documentation
- [ ] Add tests
- [ ] Implement `stream_search` that will be an enumerable that implements pagination

Also we should consider a few improvements:

- Enable users to use custom data structure (e.g. indexing and retuning a `%User{}`)
- Support more complex search queries
- Add more search options

### CODE REVIEW

File: lib/ex_aws/cloud_search/document.ex [hunk 1]
```diff
@@ -0,0 +1,5 @@
+defprotocol ExAws.CloudSearch.Document do
+  @fallback_to_any true
```
Comment (benwilson512) [inline lib/ex_aws/cloud_search/document.ex]: Curious about this choice. Are there data types other than the structs that we want to support?

File: lib/ex_aws/cloud_search/request.ex [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    {:ok, config[:json_codec].decode!(body)}
+  end
+
+  defp url(%{search_endpoint: search_endpoint}, path = "/search" <> _) do
+    Path.join(search_endpoint, Enum.join([@version, path]))
+  end
+  defp url(%{document_endpoint: document_endpoint}, path = "/documents" <> _) do
+    Path.join(document_endpoint, Enum.join([@version, path]))
+  end
+  defp url(_, _) do
```
Comment (benwilson512) [inline lib/ex_aws/cloud_search/request.ex]: Nice, I should do this more often.

File: lib/ex_aws/cloud_search/document.ex (hunk 1 above)
Comment (mschae) [inline lib/ex_aws/cloud_search/document.ex]: None, really. I guess this should rather be a behaviour right?

Comment (benwilson512) [inline lib/ex_aws/cloud_search/document.ex]: Well if there are cases where I may have some `%User{}` and want to provide some customization with respect to the way in which it's encoded, I think a protocol is fine. I'm not really all that familiar with cloudsearch yet so it isn't entirely clear if such a situation is what we want.

The particular comment here was less about the use of a protocol in and of itself, and more about whether or not we want to fall back to any vs enforce particular implementations to exist.

Comment (mschae) [inline lib/ex_aws/cloud_search/document.ex]: IC so the comment was about the `@fallback_to_any`. And you're right, we don't need it.

I will add documentation cause (a) it's nice and (b) it'll explain my choice a little.

**Comment by benwilson512** [comment]
Hey there!

I realize it's been an eon since this was created. I'm just letting you know that we have dramatically simplified the structure of ExAws as we go into 1.0. We would be happy to look at a contribution for Route53 again, but I'm gonna close this for now as most of what is here is related to boilerplate only necessary pre 1.0.

Thanks!

- Ben

**Pull request closed without merging**